File: #src/DACSDKMA/DACSDKMA-SamplePlayer-ObjectiveC/DACSDKMA-SamplePlayer-ObjectiveC/ViewController.m
//
//  ViewController.m
//  DACSDKMA-SamplePlayer-ObjectiveC
//
//  Copyright © 2016 D.A.Consortium All rights reserved.
//

#import <AVFoundation/AVFoundation.h>
#import <UIKit/UIKit.h>
#import <DACSDKMA/DACSDKMA-Swift.h>
#import "DACSDKMA_SamplePlayer_ObjectiveC-Swift.h"

#import "ViewController.h"

// サンプル動画URL
NSString *const contentUrl = @"http://vjs.zencdn.net/v/oceans.mp4";

// サンプルアドタグURL
NSString *const adTagUri = @"https://saxp.zedo.com/asw/fnsr.vast?n=2696&c=25/11&d=17&s=2&v=vast2&pu=__page-url__&ru=__referrer__&pw=__player-width__&ph=__player-height__&z=__random-number__";


@interface ViewController() <DACSDKMAAdsLoaderDelegate, DACSDKMAAdsManagerDelegate>

// 動画ビュー
@property(nonatomic, strong) UIView *videoView;
// 動画コンテンツプレイヤー
@property(nonatomic, strong) AVPlayer *contentPlayer;
// 動画再生ボタン
@property(nonatomic, strong) UIButton *playButton;

// SDK用変数
@property(nonatomic, strong) DACSDKMAAdsLoader *dacsdkmaAdsLoader;
@property(nonatomic, strong) DACSDKMAAdsManager *dacsdkmaAdsManager;
@property(nonatomic, strong) DACSDKMAAdController *dacsdkmaAdController;

@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // videoViewをセットする
    self.videoView = [[UIView alloc] initWithFrame: CGRectMake(20, 20, self.view.bounds.size.width - 40, self.view.bounds.size.height - 40)];
    self.videoView.backgroundColor = [UIColor colorWithWhite:0.0 alpha: 0.5];
    self.videoView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [self.view addSubview:self.videoView];
    
    // playButtonをセットする
    self.playButton = [[UIButton alloc] initWithFrame:self.videoView.bounds];
    [self.playButton setTitle:@"▶️" forState:UIControlStateNormal];
    [self.playButton addTarget:self action:@selector(onPlayButtonTouch:) forControlEvents:UIControlEventTouchUpInside];
    self.playButton.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.playButton.layer.zPosition = MAXFLOAT;
    [self.videoView addSubview:self.playButton];
    
    // 動画コンテンツプレイヤーを使うための準備をする。
    [self setUpContentPlayer];
    
    // SDKを使うための準備をする
    [self setupAdsLoader];
}

- (IBAction)onPlayButtonTouch:(id)sender {
    self.playButton.hidden = YES;
    [self requestAds];
}


- (void)setUpContentPlayer {
    self.contentPlayer = [AVPlayer playerWithURL:[NSURL URLWithString:contentUrl]];
    AVPlayerLayer *playerLayer = [AVPlayerLayer playerLayerWithPlayer:self.contentPlayer];
    playerLayer.frame = self.videoView.layer.bounds;
    [self.videoView.layer addSublayer:playerLayer];
}

#pragma mark SDK Setup
- (void)setupAdsLoader {
    DACSDKMASettings *settings = [[DACSDKMASettings alloc] init];
    self.dacsdkmaAdsLoader = [[DACSDKMAAdsLoader alloc] initWithSettings:settings];
    self.dacsdkmaAdsLoader.delegate = self;
}

- (void)requestAds {
    DACSDKMAAdContainer *adContainer = [[DACSDKMAAdContainer alloc] initWithView:self.videoView companionSlots:nil];
    DACSDKMAAdsRequest *request = [[DACSDKMAAdsRequest alloc] initWithAdTagURI:adTagUri adContainer: adContainer contentPlayhead: nil];
    [self.dacsdkmaAdsLoader requestAds:request];
}


#pragma mark AdsLoader Delegates
// 広告データの読み込みが正常に完了した際に呼ばれます。
- (void)dacsdkmaAdsLoader:(DACSDKMAAdsLoader *)loader didLoad:(DACSDKMAAdsLoadedData *)adsLoadedData {
    self.dacsdkmaAdsManager = adsLoadedData.adsManager;
    self.dacsdkmaAdsManager.delegate = self;
    [self.dacsdkmaAdsManager loadWithCompletion: ^(BOOL result) {
        if (result) {
            self.dacsdkmaAdController = [[DACSDKMAAdController alloc] initWithAdsManager:self.dacsdkmaAdsManager];
        }
    }];
}

// 広告データの読み込みが失敗した際に呼ばれます。
- (void)dacsdkmaAdsLoader:(DACSDKMAAdsLoader *)loader didFail:(DACSDKMAAdError *)adError {
    NSLog(@"Error: dacSdkMaAdsLoader:(loader) didFail(adError) = %@", adError.message);
    [self.contentPlayer play];
}

#pragma mark AdsManager Delegates
// DACSDKMAAdEventが発生した際に呼ばれます。
- (void)dacsdkmaAdsManager:(DACSDKMAAdsManager *)adsManager didReceiveAdEvent:(DACSDKMAAdEvent *)adEvent {
    if (adEvent.type == DACSDKMAAdEventTypeDidLoad) {
        [self.dacsdkmaAdsManager play];
    }
    else if (adEvent.type == DACSDKMAAdEventTypeDidAllAdsComplete) {
        [self.dacsdkmaAdsManager clean];
        self.dacsdkmaAdsManager = nil;
        [self.dacsdkmaAdController clean];
        self.dacsdkmaAdController = nil;
    }
}

// DACSDKMAAdErrorが発生した際に呼ばれます。
- (void)dacsdkmaAdsManager:(DACSDKMAAdsManager *)adsManager didReceiveAdError:(DACSDKMAAdError *)adError {
    NSLog(@"Error: dacSdkAdsManager:(adsManager) didReceiveAdError:(adError) = %@", adError.message);
}

// 動画広告が再生開始、レジュームした際に呼ばれます。アプリのビデオコンテンツに停止を要求します。
- (void)dacsdkmaAdsManagerDidRequestContentPause:(DACSDKMAAdsManager *)adsManager {
    [self.contentPlayer pause];
}

// 動画広告が一時停止、正常終了・エラー終了した際に呼ばれます。アプリのビデオコンテンツに再生を要求します。
- (void)dacsdkmaAdsManagerDidRequestContentResume:(DACSDKMAAdsManager *)adsManager {
    [self.contentPlayer play];
}

@end

File: #src/DACSDKMA/DACSDKMA-SampleScrollView-Swift/DACSDKMA-SampleScrollView-Swift/ViewController.swift
//
//  ViewController.swift
//  DACSDKMA-SampleScrollView-Swift
//
//  Copyright © 2015 D.A.Consortium Inc. All rights reserved.
//

import UIKit
import DACSDKMA

class ViewController: UIViewController, DACSDKMAAdsLoaderDelegate, DACSDKMAAdsManagerDelegate {

    // リクエスト先アドタグURI
    let adTagUri = "https://saxp.zedo.com/asw/fnsr.vast?n=2696&c=25/11&d=17&s=2&v=vast2&pu=__page-url__&ru=__referrer__&pw=__player-width__&ph=__player-height__&z=__random-number__"

    var scrollView: UIScrollView!
    
    // 広告掲載するView
    var adView: UIView!
    
    // SDKで使用する変数を宣言する
    var dacSdkAdsMaSettings: DACSDKMASettings?   = nil
    var dacAdsLoader: DACSDKMAAdsLoader?  = nil
    var dacAdsManager: DACSDKMAAdsManager? = nil
    var dacAdController: DACSDKMAAdController? = nil
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // self.viewのサイズを取得する
        let selfViewCGRect:     CGRect  = self.view.bounds
        let selfViewWidth:      CGFloat = selfViewCGRect.width
        let selfViewHeight:     CGFloat = selfViewCGRect.height
        
        // scrollViewを設定する
        self.scrollView = UIScrollView(frame: CGRect(x: 0, y: 0, width: selfViewWidth, height: selfViewHeight))
        self.scrollView.backgroundColor  = UIColor.blue
        self.scrollView.contentSize      = CGSize(width: selfViewWidth, height: selfViewHeight*3)
        
        // adViewを設定する
        self.adView = UIView(frame: CGRect(x: 0, y: selfViewHeight-100, width: selfViewWidth, height: 300))
        self.adView.backgroundColor = UIColor.green
        self.scrollView.addSubview(adView)
        
        // self.viewにscrollViewをaddSubviewする
        self.view.addSubview(scrollView)
        
        self.requestAds()
    }
    
    // VAST広告をリクエストする
    func requestAds() {
        self.dacSdkAdsMaSettings = DACSDKMASettings()
        self.dacAdsLoader =  DACSDKMAAdsLoader(settings: self.dacSdkAdsMaSettings!)
        self.dacAdsLoader?.delegate = self
        
        //広告動画を掲載するViewとリクエスト先アドサーバのURLをリクエストにセットする
        let adContainer: DACSDKMAAdContainer = DACSDKMAAdContainer(view: self.adView!)
        let request: DACSDKMAAdsRequest = DACSDKMAAdsRequest(adTagURI: self.adTagUri, adContainer: adContainer)
        self.dacAdsLoader?.requestAds(request)
    }
    
    // VASTレスポンスの処理が終わった時に実行される
    func dacsdkmaAdsLoader(_ loader: DACSDKMAAdsLoader, didLoad adsLoadedData: DACSDKMAAdsLoadedData) {
        self.dacAdsManager = adsLoadedData.adsManager
        self.dacAdsManager?.delegate = self
        // 広告動画ファイルのダウンロードする
        self.dacAdsManager?.load()
        
        self.dacAdController = DACSDKMAAdController(adsManager: adsLoadedData.adsManager)
    }
    
    // VASTリクエスト/レスポンスでエラーが呼ばれたときに実行される
    func dacsdkmaAdsLoader(_ loader: DACSDKMAAdsLoader, didFail adError: DACSDKMAAdError) {
        // エラーログを出力する。
        NSLog("dacSdkMaAdsLoader: didFail: \(adError.message)")
    }
    
    // AdsManagerのイベントが発生した時に実行される
    func dacsdkmaAdsManager(_ adsManager: DACSDKMAAdsManager, didReceiveAdEvent adEvent: DACSDKMAAdEvent) {
        if adEvent.type == DACSDKMAAdEventType.didLoad {
            // 広告動画ファイルがダウンロードされたら、広告動画を再生する
            adsManager.play()
        }
        else if adEvent.type == DACSDKMAAdEventType.didAllAdsComplete{
            adsManager.clean()
            self.dacAdsManager = nil
            self.dacAdController?.clean()
            self.dacAdController = nil
        }
    }
    
    // AdsManagerにエラーが発生した時に実行される
    func dacsdkmaAdsManager(_ adsManager: DACSDKMAAdsManager, didReceiveAdError adError: DACSDKMAAdError) {
        // エラーログを出力し、本編動画を再生する
        NSLog("dacSdkMaAdsManager: didReceiveAdError: \(adError.message)")
    }
    
    func dacsdkmaAdsManagerDidRequestContentPause(_ adsManager: DACSDKMAAdsManager) {
        // アプリに動画プレーヤーがないので何もしない
    }
    
    func dacsdkmaAdsManagerDidRequestContentResume(_ adsManager: DACSDKMAAdsManager) {
        // アプリに動画プレーヤーがないので何もしない
    }
}

